### ISSUE TITLE
Domain adaptation to law texts

### ISSUE THREAD

**Issue by adbar** [tsproisl/SoMaJo#21]
The syntax and abbreviations of law texts seem to confuse the tokenizer, see for example the following German example:

`(BGBl. I S. 681),` ([source](http://www.gesetze-im-internet.de/ifsg/__60.html)).

**Comment by tsproisl** [comment]
As far as I can see, this is due to specialized legal abbreviations that the tokenizer does not know, right?

I see two potential problems: Unknown abbreviations ending in a dot, e.g. *BGBl.*, and unknown dotless abbreviations that might be taken for camelCase, e.g. *AltersVersErhV* (but I guess using the camelCase option would not make a lot of sense for legal texts, anyway). Do you think that adding the relevant abbreviations, e.g. from [here](https://de.wikipedia.org/wiki/Abk%C3%BCrzungen/Gesetze_und_Recht), to the corresponding files (`abbreviations_de.txt`, `camel_case_tokens.txt`, `single_token_abbreviations_de.txt`) would be sufficient to deal with legal texts? Or am I missing something?

**Comment by adbar** [comment]
Hi @tsproisl, yes, I would say that the tokenizer doesn't find the right split because it's an unknown case.
I see two options here:
- either you add a list of abbreviations as you suggest, although exceptions lists of that kind will eventually slow down processing (or I am wrong)?
- or you use a generic regex as heuristic since combinations like _BGBl._ are rare, something like `[A-Z]{2,}[a-z]+[A-Za-z]+\.` ?

**Comment by tsproisl** [comment]
I've added the abbreviations (but not the camel case tokens) from [here](https://de.wikipedia.org/wiki/Abk%C3%BCrzungen/Gesetze_und_Recht) to `abbreviations_de.txt`, and `single_token_abbreviations_de.txt`. Actually, there weren't that many abbreviations with periods in the list. Theoretically, longer lists of abbreviations could lead to an increase in runtime (depending on how well the regular expression engine optimizes things). In practice, running a quick benchmark on a corpus of 16 million tokens, I did not notice a slowdown compared to v2.1.4.
